src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,317 @@
+        ));
+    }
+
+    let ti = res.get_target_info();
+
+    if ti.len() != SGX_TI_SIZE {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "InitQuoteEx: Invalid TARGETINFO size: {} != {}",
+                ti.len(),
+                SGX_TI_SIZE
+            ),
+        ));
+    }
+
+    out_buf.copy_from_slice(ti);
+
+    Ok(ti.len())
+}
+
+/// Gets key size
+fn get_key_size(akid: Vec<u8>) -> Result<usize, Error> {
+    let mut req = AesmRequest::new();
+    let mut msg = Request_InitQuoteExRequest::new();
+
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_b_pub_key_id(false);
+    msg.set_att_key_id(akid);
+    req.set_initQuoteExReq(msg);
+
+    let pb_msg = req.send()?;
+
+    let res = pb_msg.get_initQuoteExRes();
+
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!("InitQuoteEx error: {:?}", res.get_errorCode()),
+        ));
+    }
+
+    Ok(res.get_pub_key_id_size() as usize)
+}
+
+/// Fills the Quote obtained from the AESMD for the Report specified into
+/// the output buffer specified and returns the number of bytes written.
+fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usize, Error> {
+    if out_buf.len() != SGX_QUOTE_SIZE {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "Invalid size of output buffer {} != {}",
+                out_buf.len(),
+                SGX_QUOTE_SIZE
+            ),
+        ));
+    }
+
+    let mut req = AesmRequest::new();
+
+    let mut msg = Request_GetQuoteExRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_report(report[0..432].to_vec());
+    msg.set_att_key_id(akid);
+    msg.set_buf_size(SGX_QUOTE_SIZE as u32);
+    req.set_getQuoteExReq(msg);
+
+    let pb_msg = req.send()?;
+
+    let res = pb_msg.get_getQuoteExRes();
+
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!("GetQuoteEx error: {:?}", res.get_errorCode()),
+        ));
+    }
+
+    let quote = res.get_quote();
+
+    if quote.len() != SGX_QUOTE_SIZE {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "GetQuoteEx: Invalid QUOTE size: {} != {}",
+                quote.len(),
+                SGX_QUOTE_SIZE
+            ),
+        ));
+    }
+
+    out_buf.copy_from_slice(quote);
+    Ok(quote.len())
+}
+
+/// Returns the number of bytes written to the output buffer. Depending on
+/// whether the specified nonce is NULL, the output buffer will be filled with the
+/// Target Info for the QE, or a Quote verifying a Report.
+pub fn get_attestation(
+    nonce: usize,
+    nonce_len: usize,
+    buf: usize,
+    buf_len: usize,
+) -> Result<usize, Error> {
+    let out_buf: &mut [u8] = unsafe { from_raw_parts_mut(buf as *mut u8, buf_len) };
```

Comment:
nit;
```suggestion
    let out_buf = unsafe { from_raw_parts_mut(buf as *mut u8, buf_len) };
```
compiler should be able to infer the type here